exercises/practice/pov/src/lib.rs

```diff
@@ -0,0 +1,34 @@
+use std::fmt::Debug;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Tree<T: Clone + Debug + PartialEq> {
+    _remove_this: std::marker::PhantomData<T>,
+}
+
+impl<T: Clone + Debug + PartialEq> Tree<T> {
+    pub fn new(label: T) -> Self {
+        todo!("Implement a function that creates a new Tree with given {label:?}");
+    }
+
+    pub fn with_children(label: T, children: Vec<Self>) -> Self {
+        todo!("Implement a function that creates a new Tree with given {label:?} and {children:?}");
+    }
+
+    pub fn label(&self) -> T {
+        todo!("Implement getter for label.");
+    }
+
+    pub fn children(&self) -> Vec<&Self> {
```

Comment:
I don't like that we need this function in the first place. It's only used in the util sorter function... the tests don't require it directly. What if we just tell users that they must ensure tree comparison doesn't fail due to a different order of children? They would have to options of ensuring that:

1. Use a `HashSet` instead of a `Vec` to store children.
2. Always sort children themselves.

---

This is not relevant if we can remove the function, but I think the idiomatic API here would be:

```suggestion
pub fn children(&self) -> impl Iterator<Item = &Self> { todo!() }
```

Reply:
You mixed up the topic of the thread here. This thread is about whether or not we need users to implement the function `children` as a publicly exposed API. The comments you made about generics should be in a different comment thread. I'll repost them there in quoted blocks so we don't get things mixed up. So my answer here is only about the `children` API. Please keep threads on-topic in the future.

> All the language tracks that I've tried this exercise so far don't demand specific sort order.

We don't have to implement things exactly the same way as other language tracks. Rust has two properties that not all other langues have:
1. The implementor of a type provides its own comparison function.
1. Encapsulation is idiomatic, publicly exposing internals for the sake of testing is discouraged.

These two properties make it generally unidiomatic to extend foreign types with custom sorting logic. IMO this justifies adding some additional requirements to the exercise. Asking users to make sure their data type compares equal independent of the order of the children also fits well with the instructions themselves, which describe the trees as undirected. The operation of reparenting wouldn't make sense in a tree where the order of children matters.